

<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>C++ API &#8212; ErrorX 1.1.0 documentation</title>
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">ErrorX 1.1.0 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">C++ API</a><ul>
<li><a class="reference internal" href="#using-the-api">Using the API</a></li>
<li><a class="reference internal" href="#walkthrough">Walkthrough</a></li>
<li><a class="reference internal" href="#viewing-and-analyzing-sequencerecords">Viewing and Analyzing SequenceRecords</a></li>
<li><a class="reference internal" href="#compilation">Compilation</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/C++-API.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="c-api">
<h1>C++ API<a class="headerlink" href="#c-api" title="Permalink to this headline">¶</a></h1>
<div class="section" id="using-the-api">
<h2>Using the API<a class="headerlink" href="#using-the-api" title="Permalink to this headline">¶</a></h2>
<p>In addition to the ErrorX binary, you can integrate the functions into your own C++ application using the shared library API. The dynamic library is location in the <code class="docutils literal notranslate"><span class="pre">ErrorX/lib/</span></code> folder. In addition, all of the input files used in this tutorial can be found in the <code class="docutils literal notranslate"><span class="pre">ErrorX/documentation</span></code> folder.</p>
</div>
<div class="section" id="walkthrough">
<h2>Walkthrough<a class="headerlink" href="#walkthrough" title="Permalink to this headline">¶</a></h2>
<p>Below is a sample of code showing how you would use ErrorX in a C++ application. In this example, we are going to process a file called <code class="docutils literal notranslate"><span class="pre">test.tsv</span></code>, and output it to <code class="docutils literal notranslate"><span class="pre">test_out.tsv</span></code>. ErrorX takes TSV files with four columns: the sequence ID or name, the nucleotide sequence itself, the inferred germline sequence, and the PHRED score from the sequencer.</p>
<p>First, we need to include the relevant ErrorX headers:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#include &quot;errorx.hh&quot;</span>
<span class="c1">#include &quot;SequenceRecords.hh&quot;</span>
<span class="c1">#include &quot;ErrorXOptions.hh&quot;</span>
</pre></div>
</div>
<p>Next, we will create an ErrorXOptions object that contains all the options related to your processing. We will set the input filename to test.tsv, and the file type to TSV. ErrorX currently accepts files in FASTQ, FASTA, TSV format. We will also set the species as human - ErrorX currently supports human and mouse. Lastly, we need to set the path to the ErrorX folder. This will be used to locate different libraries needed for execution.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
</pre></div>
</div>
<p>Now, we’re ready to run the protocol. If we have provided a FASTQ or FASTA file, ErrorX will go through germline assignment and then error correction. If we use a TSV file with the germline sequence already specified, then it will only do error correction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
<p>This will run the protocol and output the results to our previously defined outfile.</p>
<p>You can also run a query using a FASTQ file, and ErrorX will do the germline assignment for you. To do so, you only need to give a FASTQ file to ErrorXOptions, and set the format type to <code class="docutils literal notranslate"><span class="pre">&quot;fastq&quot;</span></code>, as shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
<span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
</div>
<div class="section" id="viewing-and-analyzing-sequencerecords">
<h2>Viewing and Analyzing SequenceRecords<a class="headerlink" href="#viewing-and-analyzing-sequencerecords" title="Permalink to this headline">¶</a></h2>
<p>The previous example only showed how to run the protocol and output the results to a file. But, you can also directly get the results of error correction to do further analysis within your C++ application. The central object in ErrorX is a SequenceRecord - this holds all the information for one nucleotide sequence, including its germline genes, level of somatic hypermutation, chain type, and eventually its corrected sequence. There is also a SequenceRecords object, which is a collection of separate SequenceRecord objects that have been processed together. Using the SequenceRecords object we can get access to the overall error rate in our sample, individual corrected sequences, and more. Using the ErrorXOptions object that we created before, we can get a pointer to a SequenceRecords object will all the information in the output file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">SequenceRecords</span><span class="o">*</span> <span class="n">records</span> <span class="o">=</span> <span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
<p>To work with the original and corrected nucleotide sequences of one of the entries, we can access them directly through the get() method. This will return a pointer to the SequenceRecord at the index that we provide:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Original NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">full_nt_sequence</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Corrected NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">full_nt_sequence_corrected</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
</pre></div>
</div>
<p>We can also print out a summary of the data in our SequenceRecords object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">records</span><span class="o">-&gt;</span><span class="n">print_summary</span><span class="p">();</span>
</pre></div>
</div>
<p>A full working example is shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#include &quot;errorx.hh&quot;</span>
<span class="c1">#include &quot;SequenceRecords.hh&quot;</span>
<span class="c1">#include &quot;ErrorXOptions.hh&quot;</span>
<span class="c1">#include &lt;iostream&gt;</span>

<span class="nb">int</span> <span class="n">main</span><span class="p">()</span> <span class="p">{</span>
	<span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
	<span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
	
	<span class="n">errorx</span><span class="p">::</span><span class="n">SequenceRecords</span><span class="o">*</span> <span class="n">records</span> <span class="o">=</span> <span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
	<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Original NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">-&gt;</span><span class="n">full_nt_sequence</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
	<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Corrected NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">-&gt;</span><span class="n">full_nt_sequence_corrected</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
			<span class="n">records</span><span class="o">-&gt;</span><span class="n">print_summary</span><span class="p">();</span>
		<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="compilation">
<h2>Compilation<a class="headerlink" href="#compilation" title="Permalink to this headline">¶</a></h2>
<p>To use the ErrorX library in a C++ application, you only need to include two extra steps while you are compiling your code: 1. include the header files from ErrorX, and 2. dynamically link to the library. A successful clang command looks something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clang</span><span class="o">++</span> <span class="o">-</span><span class="n">std</span><span class="o">=</span><span class="n">c</span><span class="o">++</span><span class="mi">11</span> <span class="o">-</span><span class="n">IErrorX</span><span class="o">/</span><span class="n">include</span><span class="o">/</span> <span class="o">-</span><span class="n">LErrorX</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span> <span class="o">-</span><span class="n">lerrorx</span> <span class="n">ExampleApp</span><span class="o">.</span><span class="n">cc</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">-I</span></code> flag will include all of the ErrorX headers during compilation, and the <code class="docutils literal notranslate"><span class="pre">-L</span></code> and <code class="docutils literal notranslate"><span class="pre">-l</span></code> flags will makes sure the correct library is linked.</p>
<p><strong>Note:</strong> when you run the application linking to ErrorX, the ErrorX library <strong>must</strong> be in the path where your OS is going to look for dynamically linked libraries. In Linux this path is set by an environmental variable called <code class="docutils literal notranslate"><span class="pre">LD_LIBRARY_PATH</span></code>, or <code class="docutils literal notranslate"><span class="pre">DYLD_LIBRARY_PATH</span></code> on Mac. Make sure to run the following command to set the path correctly (or, even better, add this line to your <code class="docutils literal notranslate"><span class="pre">~/.bashrc</span></code> and never worry about it again!):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:ErrorX/lib/
</pre></div>
</div>
<p><a class="reference internal" href="index.html"><span class="doc">Back to Home</span></a></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">ErrorX 1.1.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Alex Sevy.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>