

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ErrorX Documentation &mdash; ErrorX 1.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> ErrorX
          

          
            
            <img src="_static/logo_horizontal_whitebackground.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">ErrorX Documentation</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#quickstart">Quickstart</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#fastq">FASTQ</a></li>
<li><a class="reference internal" href="#fasta">FASTA</a></li>
<li><a class="reference internal" href="#tsv">TSV</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
<li><a class="reference internal" href="#full-options-list">Full options list</a></li>
</ul>
</li>
<li><a class="reference internal" href="#c-api">C++ API</a><ul>
<li><a class="reference internal" href="#using-the-api">Using the API</a></li>
<li><a class="reference internal" href="#walkthrough">Walkthrough</a></li>
<li><a class="reference internal" href="#viewing-and-analyzing-sequencerecords">Viewing and Analyzing SequenceRecords</a></li>
<li><a class="reference internal" href="#compilation">Compilation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#python-api">Python API</a><ul>
<li><a class="reference internal" href="#id1">Installation</a></li>
<li><a class="reference internal" href="#id2">Walkthrough</a></li>
</ul>
</li>
<li><a class="reference internal" href="#java-api">Java API</a><ul>
<li><a class="reference internal" href="#id3">Installation</a></li>
<li><a class="reference internal" href="#id4">Walkthrough</a></li>
</ul>
</li>
<li><a class="reference internal" href="#errorxoptions">ErrorXOptions</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">ErrorX</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>ErrorX Documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="errorx-documentation">
<h1>ErrorX Documentation<a class="headerlink" href="#errorx-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>ErrorX is a software package for correction of DNA sequencing errors. In next-generation sequencing, it’s common that many of the output sequences are actually errors, different from the input DNA sequence. This can be caused by the way the DNA was prepared, amplification before sequencing, and errors introduced by the sequencing machine itself. Unfortunately, these errors can be very costly to a next-generation sequencing pipeline, as you can never be sure which sequences are true and which are mistakes.</p>
<p>ErrorX solves this problem by using deep neural networks to predict sites where an error has been introduced. Trained on a dataset of tens of millions of sequences, ErrorX can pinpoint with extremely high accuracy positions where an error was likely introduced, saving valuable time and energy by removing these bases from further analysis. In benchmark studies, ErrorX was able to remove up to 36% of errors from a dataset with a false positive rate of less than 0.05%.</p>
<p>ErrorX currently supports processing of antibody and T-cell repertoire sequencing data from human and mice, gathered on Illumina HiSeq and MiSeq instruments.</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>To install ErrorX, please visit the <a class="reference external" href="https://github.com/sevya/errorx/releases">Releases</a> page on the ErrorX Github account. Once you have downloaded the zipped archive for your operating system (this will have the extension <code class="docutils literal notranslate"><span class="pre">.tar.gz</span></code>), you are ready to unpack the folder and starting running.</p>
<p>Run the following command to extract the files from the archive:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tar</span> <span class="n">xvfz</span> <span class="n">ErrorX</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">0</span><span class="n">_mac</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>You can also use a third-party software such as <a class="reference external" href="https://www.7-zip.org/">7-zip</a> to extract the files.</p>
<p>The ErrorX binary is ready to go from there - it is located at <code class="docutils literal notranslate"><span class="pre">ErrorX/bin/errorx</span></code></p>
<p><strong>Note:</strong> the binary relies on data contained in the ErrorX folder. For best results, leave the directory structure of ErrorX intact.</p>
</div>
<div class="section" id="quickstart">
<h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h2>
<p>After installing ErrorX you are ready to run your first sequences. You can find example sequences to use as an example in the <code class="docutils literal notranslate"><span class="pre">ErrorX/documentation/</span></code> folder. Use the following command to run ErrorX prediction:</p>
<p><code class="docutils literal notranslate"><span class="pre">ErrorX/bin/errorx</span> <span class="pre">--species</span> <span class="pre">human</span> <span class="pre">--format</span> <span class="pre">fastq</span> <span class="pre">--out</span> <span class="pre">ErrorX_out.tsv</span> <span class="pre">--allow-nonproductive</span> <span class="pre">ExampleSequences.fastq</span></code></p>
<p>You will find a tab-separated file called <code class="docutils literal notranslate"><span class="pre">ErrorX_out.tsv</span></code> with all the annotated sequences and their predicted errors.</p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<div class="section" id="fastq">
<h3>FASTQ<a class="headerlink" href="#fastq" title="Permalink to this headline">¶</a></h3>
<p>The most common way to run ErrorX is with a FASTQ file. ErrorX will run germline assignment on the sequences and use sequence data and quality information to feed into the error prediction model.</p>
</div>
<div class="section" id="fasta">
<h3>FASTA<a class="headerlink" href="#fasta" title="Permalink to this headline">¶</a></h3>
<p>Alternatively, ErrorX can be run on a FASTA file. However, please note that error correction cannot be performed with FASTA input, since error correction requires the quality information that is missing in FASTA format. Running ErrorX on FASTA input will simply run germline assignment and annotate the input sequences.</p>
</div>
<div class="section" id="tsv">
<h3>TSV<a class="headerlink" href="#tsv" title="Permalink to this headline">¶</a></h3>
<p>Alternatively you can provide sequences in TSV format. If you’ve already run germline assignment with another software, you can save time and only run error prediction in ErrorX. TSV files should have four columns, separated by tabs, with no header:</p>
<ol class="simple">
<li>Sequence ID</li>
<li>Nucleotide sequence</li>
<li>Inferred germline sequence</li>
<li>PHRED score.</li>
</ol>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>The output of ErrorX is a TSV file summarizing the input sequences along with a corrected nucleotide sequence, where the predicted errors are replaced by ‘N’. If you input a FASTQ sequence, then the TSV will have information on the V, D, and J genes, as well as the level of somatic mutation and CDR3 sequence.</p>
</div>
<div class="section" id="full-options-list">
<h3>Full options list<a class="headerlink" href="#full-options-list" title="Permalink to this headline">¶</a></h3>
<p>Below is a list of all options that can be given to the application:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>-h [ --help ]					Produce help message
	
-f [ --format ] arg				Input file format. Valid entries are fastq, fasta, or tsv.
	
-o [ --out ] arg (=out.tsv)		Output file (Default=out.tsv)
	
-s [ --species ] arg (=human)	Species for IGBLAST search. Valid entries are human or mouse (Default=human)
	
--igtype arg (=Ig)				Receptor type for IGBLAST search. Valid entries are Ig or TCR. (Default=Ig)
	
-n [ --nthreads ] arg (=-1)		Number of threads to use during execution. Enter -1 to use all available (Default=-1)
	
-e [ --error-threshold ] arg (=0.730736)	Probability cutoff for a base to be considered an error. Higher=more stringent in calling errors. Don&#39;t change this value unless you know what you are doing.
	
--infile arg					Input file name
	
--version						Print ErrorX version information and exit
	
-v [ --verbose ] arg (=1)		Verbosity level: should ErrorX output extra warnings and messages? 0: don&#39;t output any message at all. 1: output progress during processing. 2: output progress and debugging messages. (default=1)
	
--allow-nonproductive			Allow nonproductive and out-of-frame sequences to be included? (default=No)
	
 --license arg					License key to activate full version of ErrorX
</pre></div>
</div>
</div>
</div>
<div class="section" id="c-api">
<h2>C++ API<a class="headerlink" href="#c-api" title="Permalink to this headline">¶</a></h2>
<div class="section" id="using-the-api">
<h3>Using the API<a class="headerlink" href="#using-the-api" title="Permalink to this headline">¶</a></h3>
<p>In addition to the ErrorX binary, you can integrate the functions into your own C++ application using the shared library API. The dynamic library is location in the <code class="docutils literal notranslate"><span class="pre">ErrorX/lib/</span></code> folder. In addition, all of the input files used in this tutorial can be found in the <code class="docutils literal notranslate"><span class="pre">ErrorX/documentation</span></code> folder.</p>
</div>
<div class="section" id="walkthrough">
<h3>Walkthrough<a class="headerlink" href="#walkthrough" title="Permalink to this headline">¶</a></h3>
<p>Below is a sample of code showing how you would use ErrorX in a C++ application. In this example, we are going to process a file called <code class="docutils literal notranslate"><span class="pre">test.tsv</span></code>, and output it to <code class="docutils literal notranslate"><span class="pre">test_out.tsv</span></code>. ErrorX takes TSV files with four columns: the sequence ID or name, the nucleotide sequence itself, the inferred germline sequence, and the PHRED score from the sequencer.</p>
<p>First, we need to include the relevant ErrorX headers:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#include &quot;errorx.hh&quot;</span>
<span class="c1">#include &quot;SequenceRecords.hh&quot;</span>
<span class="c1">#include &quot;ErrorXOptions.hh&quot;</span>
</pre></div>
</div>
<p>Next, we will create an ErrorXOptions object that contains all the options related to your processing. We will set the input filename to test.tsv, and the file type to TSV. ErrorX currently accepts files in FASTQ, FASTA, TSV format. We will also set the species as human - ErrorX currently supports human and mouse. Lastly, we need to set the path to the ErrorX folder. This will be used to locate different libraries needed for execution.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
</pre></div>
</div>
<p>Now, we’re ready to run the protocol. If we have provided a FASTQ or FASTA file, ErrorX will go through germline assignment and then error correction. If we use a TSV file with the germline sequence already specified, then it will only do error correction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
<p>This will run the protocol and output the results to our previously defined outfile.</p>
<p>You can also run a query using a FASTQ file, and ErrorX will do the germline assignment for you. To do so, you only need to give a FASTQ file to ErrorXOptions, and set the format type to <code class="docutils literal notranslate"><span class="pre">&quot;fastq&quot;</span></code>, as shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
<span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
</div>
<div class="section" id="viewing-and-analyzing-sequencerecords">
<h3>Viewing and Analyzing SequenceRecords<a class="headerlink" href="#viewing-and-analyzing-sequencerecords" title="Permalink to this headline">¶</a></h3>
<p>The previous example only showed how to run the protocol and output the results to a file. But, you can also directly get the results of error correction to do further analysis within your C++ application. The central object in ErrorX is a SequenceRecord - this holds all the information for one nucleotide sequence, including its germline genes, level of somatic hypermutation, chain type, and eventually its corrected sequence. There is also a SequenceRecords object, which is a collection of separate SequenceRecord objects that have been processed together. Using the SequenceRecords object we can get access to the overall error rate in our sample, individual corrected sequences, and more. Using the ErrorXOptions object that we created before, we can get a pointer to a SequenceRecords object will all the information in the output file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">errorx</span><span class="p">::</span><span class="n">SequenceRecords</span><span class="o">*</span> <span class="n">records</span> <span class="o">=</span> <span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
<p>To work with the original and corrected nucleotide sequences of one of the entries, we can access them directly through the get() method. This will return a pointer to the SequenceRecord at the index that we provide:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Original NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">full_nt_sequence</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Corrected NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">full_nt_sequence_corrected</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
</pre></div>
</div>
<p>We can also print out a summary of the data in our SequenceRecords object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">records</span><span class="o">-&gt;</span><span class="n">print_summary</span><span class="p">();</span>
</pre></div>
</div>
<p>A full working example is shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#include &quot;errorx.hh&quot;</span>
<span class="c1">#include &quot;SequenceRecords.hh&quot;</span>
<span class="c1">#include &quot;ErrorXOptions.hh&quot;</span>
<span class="c1">#include &lt;iostream&gt;</span>

<span class="nb">int</span> <span class="n">main</span><span class="p">()</span> <span class="p">{</span>
	<span class="n">errorx</span><span class="p">::</span><span class="n">ErrorXOptions</span> <span class="n">options</span><span class="p">(</span> <span class="s2">&quot;test.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;test_out.tsv&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;human&quot;</span> <span class="p">);</span>
	<span class="n">options</span><span class="o">.</span><span class="n">errorx_base</span><span class="p">(</span> <span class="s2">&quot;/path/to/ErrorX/&quot;</span> <span class="p">);</span>
	<span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol_write</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
	
	<span class="n">errorx</span><span class="p">::</span><span class="n">SequenceRecords</span><span class="o">*</span> <span class="n">records</span> <span class="o">=</span> <span class="n">errorx</span><span class="p">::</span><span class="n">run_protocol</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
	<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Original NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">-&gt;</span><span class="n">full_nt_sequence</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
	<span class="n">std</span><span class="p">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s2">&quot;Corrected NT sequence: &quot;</span> <span class="o">&lt;&lt;</span> <span class="n">records</span><span class="o">-&gt;</span><span class="n">get</span><span class="p">(</span> <span class="mi">0</span> <span class="p">)</span><span class="o">-&gt;</span><span class="n">full_nt_sequence_corrected</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="n">std</span><span class="p">::</span><span class="n">endl</span><span class="p">;</span>
			<span class="n">records</span><span class="o">-&gt;</span><span class="n">print_summary</span><span class="p">();</span>
		<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="compilation">
<h3>Compilation<a class="headerlink" href="#compilation" title="Permalink to this headline">¶</a></h3>
<p>To use the ErrorX library in a C++ application, you only need to include two extra steps while you are compiling your code: 1. include the header files from ErrorX, and 2. dynamically link to the library. A successful clang command looks something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clang</span><span class="o">++</span> <span class="o">-</span><span class="n">std</span><span class="o">=</span><span class="n">c</span><span class="o">++</span><span class="mi">11</span> <span class="o">-</span><span class="n">IErrorX</span><span class="o">/</span><span class="n">include</span><span class="o">/</span> <span class="o">-</span><span class="n">LErrorX</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span> <span class="o">-</span><span class="n">lerrorx</span> <span class="n">ExampleApp</span><span class="o">.</span><span class="n">cc</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">-I</span></code> flag will include all of the ErrorX headers during compilation, and the <code class="docutils literal notranslate"><span class="pre">-L</span></code> and <code class="docutils literal notranslate"><span class="pre">-l</span></code> flags will makes sure the correct library is linked.</p>
<p><strong>Note:</strong> when you run the application linking to ErrorX, the ErrorX library <strong>must</strong> be in the path where your OS is going to look for dynamically linked libraries. In Linux this path is set by an environmental variable called <code class="docutils literal notranslate"><span class="pre">LD_LIBRARY_PATH</span></code>, or <code class="docutils literal notranslate"><span class="pre">DYLD_LIBRARY_PATH</span></code> on Mac. Make sure to run the following command to set the path correctly (or, even better, add this line to your <code class="docutils literal notranslate"><span class="pre">~/.bashrc</span></code> and never worry about it again!):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:ErrorX/lib/
</pre></div>
</div>
</div>
</div>
<div class="section" id="python-api">
<h2>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Installation<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>The Python packages are also located in the folder you just extracted. ErrorX supports Python versions 2.7 and 3.6. To install the Python 2.7 package simply run</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">python2_bindings</span><span class="o">/</span>
</pre></div>
</div>
<p>and for version 3.6 run</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">python3_bindings</span><span class="o">/</span>
</pre></div>
</div>
<p>You are now ready to run ErrorX through the Python interface!</p>
</div>
<div class="section" id="id2">
<h3>Walkthrough<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>To run the Python interface, you first need to import the errorx package to have access to all the functions to do data correction. You can also import the ErrorXOptions class if you want to have more fine-tuned control over the settings. To run the entire protocol, you need to provide an input file, as well as its format. A full example is provided in <code class="docutils literal notranslate"><span class="pre">ErrorX/documentation/ExampleApp.py</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>import errorx as ex
from errorx import ErrorXOptions

options = ErrorXOptions(&#39;example.fastq&#39;,&#39;fastq&#39;)
options.outfile( &#39;example_out.tsv&#39; )
options.species( &#39;mouse&#39; )
options.igtype( ‘Ig’ )
ex.run_protocol( options )
</pre></div>
</div>
<p>You can also get corrected sequences from within the Python script that you can then operate on, shown below. This function takes in lists of nucleotide sequences, their inferred germline sequences, and the PHRED quality scores, and returns a list of corrected sequences corresponding to the input sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;TACTCCCG&#39;</span><span class="p">]</span> 
<span class="n">germline_sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;TACTCCCA&#39;</span><span class="p">]</span>
<span class="n">phred_scores</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;HEIHHIII&#39;</span><span class="p">]</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">correct_sequences</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">germline_sequences</span><span class="p">,</span> <span class="n">phred_scores</span><span class="p">,</span> <span class="n">options</span> <span class="p">)</span>
</pre></div>
</div>
<p>You can also input a single sequence, and get the predicted probability of error for each position along that sequence (positions are zero-indexed):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;TACTCCCG&#39;</span>
<span class="n">germline_sequence</span> <span class="o">=</span> <span class="s1">&#39;TACTCCCA&#39;</span>
<span class="n">phred_scores</span> <span class="o">=</span> <span class="s1">&#39;HEIHHIII&#39;</span>

<span class="n">probabilities</span> <span class="o">=</span> <span class="n">get_predicted_errors</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">germline_sequences</span><span class="p">,</span> <span class="n">phred_scores</span><span class="p">,</span> <span class="n">options</span> <span class="p">)</span>

<span class="c1"># returns probability of error at position 0</span>
<span class="nb">print</span> <span class="n">probabilities</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> 
</pre></div>
</div>
</div>
</div>
<div class="section" id="java-api">
<h2>Java API<a class="headerlink" href="#java-api" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id3">
<h3>Installation<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>The ErrorX Java library is also located in the gzipped archive, in the folder java_bindings. To install you don’t need to do anything special - all the functionality is contained in the <code class="docutils literal notranslate"><span class="pre">ErrorX.jar</span></code> archive within the <code class="docutils literal notranslate"><span class="pre">java_bindings</span></code> folder. When you are compiling your Java application simply add the flag</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">cp</span> <span class="s1">&#39;ErrorX/java_bindings/ErrorX.jar&#39;</span>
</pre></div>
</div>
<p>to make sure the JAR is linked, and you are all set!</p>
<p>Note: the JAR relies on data files contained in the java_bindings folder. For best results, leave the directory structure of java_bindings intact and only link the JAR. Alternatively, if you have to move the JAR file, make sure to move all the other directories in <code class="docutils literal notranslate"><span class="pre">java_bindings</span></code> along with it.</p>
</div>
<div class="section" id="id4">
<h3>Walkthrough<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>Similar to the Python interface, the Java API consists of three main functions:</p>
<ol class="simple">
<li>Runs the whole protocol on a FASTQ or TSV file and writes to another TSV file,</li>
<li>Takes in an array of sequences, germline sequences, and PHRED scores and returns an array of corrected sequences,</li>
<li>Takes in a single sequence, germline sequence, and PHRED score and returns an array where element i is the probability that nucleotide at position i is an error (positions are zero-indexed).</li>
</ol>
<p>Below are examples of how to run each scenario. A full example is presented in <code class="docutils literal notranslate"><span class="pre">ErrorX/documentation/ExampleApp.java</span></code>. First import the errorx package.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">errorx.ErrorX</span><span class="p">;</span>
</pre></div>
</div>
<p>Next, in the body of your class, you need to create an ErrorXOptions object that defines, at minimum, the name of the file to be processed and its format (either FASTQ or TSV). ErrorXOptions can also be used for more advanced control of the protocol.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ErrorXOptions</span> <span class="n">options</span> <span class="o">=</span> <span class="n">new</span> <span class="n">ErrorXOptions</span><span class="p">(</span> <span class="s2">&quot;example.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">outfile</span><span class="p">(</span> <span class="s2">&quot;out.tsv&quot;</span> <span class="p">);</span>
<span class="n">options</span><span class="o">.</span><span class="n">species</span><span class="p">(</span> <span class="s2">&quot;mouse&quot;</span> <span class="p">);</span>
	
<span class="n">ErrorX</span> <span class="n">ex</span> <span class="o">=</span> <span class="n">new</span> <span class="n">ErrorX</span><span class="p">();</span>
<span class="n">ex</span><span class="o">.</span><span class="n">runProtocol</span><span class="p">(</span> <span class="n">options</span> <span class="p">);</span>
</pre></div>
</div>
<p>For scenario two, you need to make arrays of Strings representing the nucleotide sequence, germline sequence, and PHRED score of your query. The function <code class="docutils literal notranslate"><span class="pre">correctSequences</span></code> will return an array of Strings with the corrected nucleotide sequences corresponding to the input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">String</span> <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;TACTCCCG&quot;</span><span class="p">;</span>
<span class="n">String</span> <span class="n">germline_sequence</span> <span class="o">=</span> <span class="s2">&quot;TACTCCCA&quot;</span><span class="p">;</span>
<span class="n">String</span> <span class="n">phred_score</span> <span class="o">=</span> <span class="s2">&quot;HEIHHIII&quot;</span><span class="p">;</span>

<span class="n">String</span><span class="p">[]</span> <span class="n">sequences</span> <span class="o">=</span> <span class="n">new</span> <span class="n">String</span><span class="p">[</span><span class="mi">1</span><span class="p">];</span>
<span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">;</span>

<span class="n">String</span><span class="p">[]</span> <span class="n">germline_sequences</span> <span class="o">=</span> <span class="n">new</span> <span class="n">String</span><span class="p">[</span><span class="mi">1</span><span class="p">];</span>
<span class="n">germline_sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">germline_sequence</span><span class="p">;</span>

<span class="n">String</span><span class="p">[]</span> <span class="n">phred_scores</span> <span class="o">=</span> <span class="n">new</span> <span class="n">String</span><span class="p">[</span><span class="mi">1</span><span class="p">];</span>
<span class="n">phred_scores</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">phred_score</span><span class="p">;</span>

<span class="n">String</span><span class="p">[]</span> <span class="n">correctedSequences</span> <span class="o">=</span> <span class="n">new</span> <span class="n">ErrorX</span><span class="p">()</span><span class="o">.</span><span class="n">correctSequences</span><span class="p">(</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">germline_sequences</span><span class="p">,</span> <span class="n">phred_scores</span> <span class="p">);</span>
</pre></div>
</div>
<p>For the final scenario, you need to input a single sequence, germline, and PHRED score. The function <code class="docutils literal notranslate"><span class="pre">getPredictedErrors</span></code> will return an array of doubles representing the probability of error at each position (zero-indexed):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">double</span><span class="p">[]</span> <span class="n">errorPredictions</span> <span class="o">=</span> <span class="n">new</span> <span class="n">ErrorX</span><span class="p">()</span><span class="o">.</span><span class="n">getPredictedErrors</span><span class="p">(</span>
<span class="n">sequence</span><span class="p">,</span> <span class="n">germline_sequence</span><span class="p">,</span> <span class="n">phred_score</span> <span class="p">);</span>

<span class="n">System</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">println</span><span class="p">(</span> <span class="n">errorPredictions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="p">);</span> <span class="o">//</span> <span class="n">prints</span> <span class="n">the</span> <span class="n">probability</span> <span class="n">of</span> <span class="n">error</span> <span class="n">at</span> <span class="n">position</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="errorxoptions">
<h2>ErrorXOptions<a class="headerlink" href="#errorxoptions" title="Permalink to this headline">¶</a></h2>
<p>All the options surrounding the ErrorX protocol are controlled by a class called ErrorXOptions. This has the same function in the C++, Python, and Java APIs. The options that you can control with this class are the same as the command-line options. Please see the documentation on full options list above for full details on which options can be controlled.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Alex Sevy

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(false);
      });
  </script>

  
  
    
   

</body>
</html>